share/vss/vss.go

```diff
@@ -0,0 +1,630 @@
+// run.
+func (v *Verifier) Key() (abstract.Scalar, abstract.Point) {
+	return v.longterm, v.pub
+}
+
+// Index returns the index of the verifier in the list of participants used
+// during this run of the protocol.
+func (v *Verifier) Index() int {
+	return v.index
+}
+
+// SessionID returns the session id generated by the Dealer. WARNING: it returns
+// an nil slice if the verifier has not received the Deal yet !
+func (v *Verifier) SessionID() []byte {
+	return v.sid
+}
+
+// RecoverSecret recovers the secret shared by a Dealer by gathering at least t
+// Deals from the verifiers. It returns an error if there is not enough Deals or
+// if all Deals don't have the same SessionID.
+func RecoverSecret(suite abstract.Suite, deals []*Deal, n, t int) (abstract.Scalar, error) {
+	shares := make([]*share.PriShare, len(deals))
+	for i, deal := range deals {
+		// all sids the same
+		if bytes.Equal(deal.SessionID, deals[0].SessionID) {
+			shares[i] = deal.SecShare
+		} else {
+			return nil, errors.New("vss: all deals need to have same session id")
+		}
+	}
+	return share.RecoverSecret(suite, shares, t, n)
+}
+
+// aggregator is used to collect all deals, responses and justification for one
+// protocol run. It brings common functionalities for both Dealer and Verifier
+// structs.
+type aggregator struct {
+	suite     abstract.Suite
+	dealer    abstract.Point
+	verifiers []abstract.Point
+	commits   []abstract.Point
+
+	responses map[uint32]*Response
+	sid       []byte
+	deal      *Deal
+	t         int
+	badDealer bool
+}
+
+func newAggregator(suite abstract.Suite, dealer abstract.Point, verifiers, commitments []abstract.Point, t int, sid []byte) *aggregator {
+	agg := &aggregator{
+		suite:     suite,
+		dealer:    dealer,
+		verifiers: verifiers,
+		commits:   commitments,
+		t:         t,
+		sid:       sid,
+		responses: make(map[uint32]*Response),
+	}
+	return agg
+}
+
+// used only by the Dealer
+func (a *aggregator) setDeal(d *Deal) {
+	a.deal = d
+}
+
+var errDealAlreadyProcessed = errors.New("vss: verifier already received a deal")
+
+// VerifyDeal analyzes the deal and returns an error if it's incorrect. If
+// inclusion is true, it also returns an error if it the second time this struct
+// analyzes a Deal.
+func (a *aggregator) VerifyDeal(d *Deal, inclusion bool) error {
+	if a.deal != nil && inclusion {
+		return errDealAlreadyProcessed
+
+	}
+	if a.deal == nil {
+		a.commits = d.Commitments
+		a.sid = d.SessionID
+		a.deal = d
+	}
+
+	if !validT(int(d.T), a.verifiers) {
+		return errors.New("vss: invalid t received in Deal")
+	}
+
+	if !bytes.Equal(a.sid, d.SessionID) {
+		return errors.New("vss: find different sessionIDs from Deal")
+	}
+
+	if err := sign.VerifySchnorr(a.suite, a.dealer, msgDeal(d), d.Signature); err != nil {
+		return err
+	}
+
+	fi := d.SecShare
+	gi := d.RndShare
+	if fi.I != gi.I {
+		return errors.New("vss: not the same index for f and g share in Deal")
+	} else if fi.I < 0 || fi.I >= len(a.verifiers) {
```

Comment:
style: I would remove the `else` and put the `if` on the next line - it is easier to read than the `} else if fi`...